### ISSUE TITLE
Requesting a login code returns HTTP 400

### ISSUE THREAD

**Issue by micheljung** [ory/kratos#4052]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

When Requesting a code via e-mail for passwordless login, the API returns HTTP 400, but I expect it to return 200 because the request is valid.

### Reproducing the bug

Create a new account:

```shell
export KRATOS_URL=http://localhost:4433
export EMAIL=user@example.com
export PASSWORD=iohuasf0897zAJHf

actionUrl=$(\
  curl -s -X GET -H "Accept: application/json" \
    "$KRATOS_URL/self-service/registration/api" \
    | sed -n 's/.*"action":"\([^"]*\)".*/\1/p'
)

curl -s -X POST -H  "Accept: application/json" -H "Content-Type: application/json" \
  -d '{"method": "password", "password": "'$PASSWORD'", "traits": {"email": "'$EMAIL'"}}' \
  "$actionUrl"
```

Request a login code via E-Mail:

```sh
export KRATOS_URL=http://localhost:4433
export EMAIL=user@example.com

actionUrl=$(\
  curl -X GET -H "Accept: application/json" \
    "$KRATOS_URL/self-service/login/api" \
    | sed -n 's/.*"action":"\([^"]*\)".*/\1/p'
)
  
curl -i -X POST -H  "Accept: application/json" -H "Content-Type: application/json" \
-d '{"identifier": "'$EMAIL'", "method": "code"}' \
"$actionUrl" 
```

### Relevant log output

```shell
HTTP/1.1 400 Bad Request
Cache-Control: private, no-cache, no-store, must-revalidate
Content-Type: application/json; charset=utf-8
Vary: Origin
Vary: Cookie
Date: Thu, 22 Aug 2024 10:09:49 GMT
Content-Length: 1988

{"id":"367c97cc-0499-47ec-9a3c-208d2c308445","organization_id":null,"type":"api","expires_at":"2024-08-22T10:19:49.24531247Z","issued_at":"2024-08-22T10:09:49.24531247Z","request_url":"http://localhost:4433/self-service/login/api","active":"code","ui":{"action":"http://127.0.0.1:4433/self-service/login?flow=367c97cc-0499-47ec-9a3c-208d2c308445","method":"POST","nodes":[{"type":"input","group":"default","attributes":{"name":"csrf_token","type":"hidden","value":"","required":true,"disabled":false,"node_type":"input"},"messages":[],"meta":{}},{"type":"input","group":"default","attributes":{"name":"identifier","type":"hidden","value":"user@example.com","required":true,"disabled":false,"node_type":"input"},"messages":[],"meta":{"label":{"id":1070002,"text":"E-Mail","type":"info","context":{"title":"E-Mail"}}}},{"type":"input","group":"code","attributes":{"name":"method","type":"hidden","value":"code","disabled":false,"node_type":"input"},"messages":[],"meta":{}},{"type":"input","group":"code","attributes":{"name":"code","type":"text","required":true,"disabled":false,"node_type":"input"},"messages":[],"meta":{"label":{"id":1070013,"text":"Login code","type":"info"}}},{"type":"input","group":"code","attributes":{"name":"method","type":"submit","value":"code","disabled":false,"node_type":"input"},"messages":[],"meta":{"label":{"id":1070005,"text":"Submit","type":"info"}}},{"type":"input","group":"code","attributes":{"name":"resend","type":"submit","value":"code","disabled":false,"node_type":"input"},"messages":[],"meta":{"label":{"id":1070008,"text":"Resend code","type":"info"}}}],"messages":[{"id":1010014,"text":"An email containing a code has been sent to the email address you provided. If you have not received an email, check the spelling of the address and retry the login.","type":"info"}]},"created_at":"2024-08-22T10:09:49.251261Z","updated_at":"2024-08-22T10:09:49.251261Z","refresh":false,"requested_aal":"aal1","state":"sent_email","transient_payload":{}}
```

### Relevant configuration

```yml
methods:
    password:
      enabled: true
    totp:
      config:
        issuer: Kratos
      enabled: false
    lookup_secret:
      enabled: false
    link:
      enabled: false
    code:
      passwordless_enabled: true
    passkey:
      enabled: true
      config:
        rp:
          display_name: My App
          id: localhost
          origins:
            - http://127.0.0.1:4455
```

### Version

v1.2.0

### On which operating system are you observing this issue?

Linux

### In which environment are you deploying?

Docker Compose

**Comment by jkoppolu** [comment]
We are facing the same issue with Browser API. 
When we update the login flow to trigger the login code, everything works fine—the Code is generated, an email is triggered, and the API response looks good. It is just that the HTTP status is 400. 

`POST /self-service/login?flow=2f3c2b78-e7ef-4b9a-b22b-d4cd1117444f HTTP/1.1
X-Session-Token: <string>
Cookie: <string>; csrf_token_82b119fa58a0a1cb6faa9738c1d0dbbf04fcc89a657b7beb31fcde400ced48ab=hj1chWyh2KjVr1RPqvkpr+H1AYSIFcmQOpr7O9YFqnw=
Content-Type: application/x-www-form-urlencoded
Accept: application/json
User-Agent: PostmanRuntime/7.42.0
Cache-Control: no-cache
Postman-Token: 73526f77-cc82-4224-902c-96f6d7e870a5
Accept-Encoding: gzip, deflate, br
Connection: keep-alive
Content-Length: 154
Referer: http://localhost:4434/admin/self-service/login?flow=2f3c2b78-e7ef-4b9a-b22b-d4cd1117444f
Host: localhost:4433
 
method=code&identifier=test%40gmail.com&csrf_token=KOA2s4ZH7zGeF1NUJlvfahJqX3aMG8qcE%2BQNUursuG6u3Wo26uY3mUu4BxuMovbF859e8gQOAwwpfvZpPOkSEg%3D%3D
 
HTTP/1.1 400 Bad Request
Cache-Control: private, no-cache, no-store, must-revalidate
Content-Type: application/json; charset=utf-8
Vary: Origin
Vary: Cookie
Date: Mon, 16 Sep 2024 07:08:05 GMT
Transfer-Encoding: chunked
 
{
  "id": "2f3c2b78-e7ef-4b9a-b22b-d4cd1117444f",
  "organization_id": null,
  "type": "browser",
  "expires_at": "2024-09-16T07:17:02.384872Z",
  "issued_at": "2024-09-16T07:07:02.384872Z",
  "request_url": "http://localhost:4433/self-service/login/browser",
  "active": "code",
  "ui": {
    "action": "http://localhost:4433/self-service/login?flow=2f3c2b78-e7ef-4b9a-b22b-d4cd1117444f",
    "method": "POST",
    "nodes": [
      {
        "type": "input",
        "group": "default",
        "attributes": {
          "name": "csrf_token",
          "type": "hidden",
          "value": "JAsx7gK1Zewg12kEW5am3GsArsi57Gdr1w5ISyPD5smiNm1rbhS9RPV4PUvxb49zivWvTDH5rvvtlLNw9cZMtQ==",
          "required": true,
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {}
      },
      {
        "type": "input",
        "group": "default",
        "attributes": {
          "name": "identifier",
          "type": "hidden",
          "value": "test@gmail.com",
          "required": true,
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": {
            "id": 1070002,
            "text": "E-Mail",
            "type": "info",
            "context": { "title": "E-Mail" }
          }
        }
      },
      {
        "type": "input",
        "group": "code",
        "attributes": {
          "name": "method",
          "type": "hidden",
          "value": "code",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {}
      },
      {
        "type": "input",
        "group": "code",
        "attributes": {
          "name": "code",
          "type": "text",
          "required": true,
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": { "id": 1070013, "text": "Login code", "type": "info" }
        }
      },
      {
        "type": "input",
        "group": "code",
        "attributes": {
          "name": "method",
          "type": "submit",
          "value": "code",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": { "label": { "id": 1070005, "text": "Submit", "type": "info" } }
      },
      {
        "type": "input",
        "group": "code",
        "attributes": {
          "name": "resend",
          "type": "submit",
          "value": "code",
          "disabled": false,
          "node_type": "input"
        },
        "messages": [],
        "meta": {
          "label": { "id": 1070008, "text": "Resend code", "type": "info" }
        }
      }
    ],
    "messages": [
      {
        "id": 1010014,
        "text": "An email containing a code has been sent to the email address you provided. If you have not received an email, check the spelling of the address and retry the login.",
        "type": "info"
      }
    ]
  },
  "created_at": "2024-09-16T12:37:02.391149Z",
  "updated_at": "2024-09-16T12:37:02.391149Z",
  "refresh": false,
  "requested_aal": "aal1",
  "state": "sent_email",
  "transient_payload": {}
}
`

**Comment by aeneasr** [comment]
Expected behavior

**Comment by jkoppolu** [comment]
Thanks, @aeneasr, for confirming the behavior. 

400 response, in this case, sends a wrong signal that the server could not process the request even though it successfully did.  Just FYI. The issue is raised here, too -  https://github.com/ory/kratos/issues/3796

**Comment by aeneasr** [comment]
200 OK could be misinterpreted as the login succeeded, which it did not. The server failed to validate the response (invalid credentials), hence an error code is to be expected. We're not going to change this behavior, it would be a huge breaking change for thousands of adopters and customers

**Comment by aeneasr** [comment]
To me it sounds like you're facing an issue in the SDK (or something else) that you find frustrating. Arguing over HTTP status codes is not going to really address that issue. So what exactly is your issue, happy to take a look at that, but we're not going to spend cycles on arguing over something that isn't going to change, given the information we have today.

**Comment by micheljung** [comment]
I'm not facing an issue with the SDK but with the API.

What's frustrating is that the API returns an HTTP 400 for a valid request, which is very obviously wrong, and that this issue is disregarded with an argument that doesn't hold up - even though there are already 3 developers reporting this on GitHub - with the fear that someone *might* interpret HTTP 200 as "successful login" even though it's 100% clear that the request requests a login code and will never result in a successful login, so nobody would ever expect that it would.

The issue is that we talk to the API and the API gives us misleading status codes, so we can't rely on the status codes to detect whether our request was successful or not, like you normally do for any HTTP API that properly implements status codes.

Unless the API is just an implementation detail of the SDK and not to be used directly, I don't see why we need to bring the SDK into this.

**Comment by aeneasr** [comment]
Which one of these is a failed login attempt, and which one is a successful?

```
POST /self-service/login
400 Bad Request
200 OK
```

Now if we return 200, can you tell them apart?

```
POST /self-service/login
200 OK
200 OK
```

You can't! Instead you now need to interpret the body. But which field? Will it ever change in the API? What if the body can't be read for performance reasons?

Granted, in this specific case the submit is successful and it's a bit weird, but in other cases ("user does not exist") it's obvious that we want to be as explicit as we can about the status of the login.

Because we do this in login, we apply the same principle everywhere else, to stay consistent.

**Comment by micheljung** [comment]
I'm happy to read that you think that having to process the response body is bad. Because this is exactly what this issue is about :)

Your example doesn't tell the whole story as POST requests are not equal simply because they target the same endpoint. Each POST request has a body and it's that body that specifies what exactly was requested.

If I send a "request login code" request and get HTTP 200, I know which request worked and I don't need to process the body.
If I send a "submit login code" request and get HTTP 200, I know which request worked and I don't need to process the body.

Because every response is associated with a known request, everyone can tell them apart with 100% accuracy and there is no need to process the body.

In the current situation, however, you'll *always* get HTTP 400 for the first request. How do you know if the request was successful or not? To describe this in your own words:

> How can you tell them apart? You can't! Instead, you now need to interpret the body. But which field? Will it ever change in the API? What if the body can't be read for performance reasons?

Right now, we are **forced** to process the body every single time. There is no other way to find out whether the request was successful.

If you think processing the body is bad, like you said you do, then it should be in your best interest to fix this issue.

(btw. the correct HTTP status code for a login error due to incorrect credentials is 401 Unauthorized.)

**Comment by micheljung** [comment]
@aeneasr what are your thoughts?

To give you another example of how this causes pain: The most basic metrics that are often collected are HTTP status codes. This helps to identify how frequently your users receive errors, and detect potential issues. HTTP 400 is classified as an error. The login endpoint produces thousands of false errors that appear in the metrics, killing your signal-to-noise ratio and thus rendering your metrics useless.

To sum it up, this issue:

* Violates [RFC 7231 6.5.1](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.1)
* Confuses developers
* Forces them to implement a workaround by parsing the request body, which you agree is bad
* Makes client code brittle
* Costs performance
* Screws up your metrics

So what exactly is the argument that weighs heavier than these problems?

**Comment by aeneasr** [comment]
To be truthful, I think most of the points made are just a matter of mindset. If you raise an alert for a 400 code, sure, you'll get a lot of alerts. But if you know that 400 isn't a bad thing, don't raise an alert. Software doesn't prevent you from doing this.

I did however do some more research - like any modern dev would do - using ChatGPT (a great tool to this task because we query the internet's "compressed" knowledge). I think it had some good points:

---

What response code should a JSON API give when there are form validation errors?

> A JSON API should return a `422 Unprocessable Entity` response code when there are form validation errors. This status code indicates that the server understands the content type of the request and the syntax is correct, but the data provided contains semantic errors, such as invalid input fields.

And what status code if the flow is interrupted with a message (for example because an email was sent with an activation link)?

> In cases where the flow is interrupted with a message, such as sending an email with an activation link, the API should return a `202 Accepted` status code. This indicates that the request has been received and understood, but further processing will take place asynchronously, such as waiting for the user to activate their account via the email link.

What if we need to perform a second step (for example login, then multi factor)?

> If a second step is required, such as logging in followed by multi-factor authentication (MFA), the API should return a `401 Unauthorized` status code with additional information in the response body, such as instructions or a URL for the next step. The `401` status code informs the client that further authentication steps are necessary before granting access. 
>
> Alternatively, you could use a `200 OK` status with a clear response body that indicates the need for further authentication if you want to avoid using `401`. However, `401` is typically more appropriate for multi-step authentication flows.

When should 200 OK be sent?

> The `200 OK` status code should be sent when the request is successfully processed, and no further action or steps are required from the client. It indicates that the request was received, understood, and handled properly, with a successful outcome. Here are common situations when to use `200 OK`:
>
>1. **Fetching resources**: When a GET request retrieves data successfully.
>2. **Successful submission**: When a POST request successfully creates or submits data without any additional steps required.
>3. **Completing a flow**: When an action, such as authentication or form submission, is fully completed without errors or further interaction needed.
>
>In cases where a second step or additional action is needed (e.g., MFA or email verification), you would typically use other status codes like `401 Unauthorized` or `202 Accepted` rather than `200 OK`.

To summarize:

>- **422 Unprocessable Entity**: Use when there are form validation errors.
>- **202 Accepted**: Use when the flow is interrupted, such as sending an email with an activation link.
>- **401 Unauthorized**: Use when additional authentication steps are required, like MFA after login.
>- **200 OK**: Use when the request is successfully processed, and no further action is needed.

Should we replace 400 bad request with more specific codes?

> You should change the status codes from `400 Bad Request` to more specific ones like `422`, `202`, and `401` where appropriate. While `400` is technically valid for general client-side errors, it does not provide the same level of detail and clarity as using the correct status codes:
>
>- **422** is more specific for validation errors, indicating the request was syntactically correct but semantically incorrect.
>- **202** conveys that the request was accepted but will be processed later, which gives clearer feedback in cases of interrupted flows like sending activation emails.
>- **401** clearly signals the need for further authentication steps, such as MFA.
>
>Using specific status codes helps clients understand the exact nature of the issue and respond accordingly, leading to better error handling and user experience.
>


>Should I instead replace everything with 200 OK and have the state in the body?
>
>It is not recommended to replace everything with `200 OK` and handle the state in the body. While it is technically possible, this approach reduces clarity and violates standard API practices. Here are the key reasons:
>
>1. **Separation of concerns**: HTTP status codes provide a clear, standardized way to convey the outcome of a request, while the response body contains detailed data. Mixing these concerns makes it harder for clients to differentiate between successful, pending, or error states. 
>2. **Client-side handling**: Many clients, libraries, and tools rely on HTTP status codes to manage responses (e.g., retrying on `401 Unauthorized`). Using only `200 OK` forces clients to parse the body for every response, increasing complexity.
>
>3. **Error reporting**: Status codes like `422` and `401` immediately inform the client what went wrong (validation errors, authentication steps). Relying solely on the body for this information adds unnecessary complexity and may result in less efficient error handling.
>
>Instead, follow best practices by using the appropriate status codes (`422`, `202`, `401`, etc.), while using the response body to provide additional context or details. This creates clearer, more maintainable APIs.

---

So I would summarize it like this. We have opportunity to improve status code for certain endpoints (but need to do it without breaking changes to existing clients which will be very challenging) and their responses. Sending 200 OK for anything where the request is not successful and finished is not in the mix.

We will probably consider status codes once we work on some other refactoring around the flows, but that's realistically not going to happen prior to 2026 ...

**Comment by micheljung** [comment]
Thank you for taking the time to look into this.

If you're using ChatGPT as an argument, here is my take:


I'm writing an identity server that provides login via an email code. To log in, a user has to send a POST request containing their e-mail address to /self-service/login/api. This will send an email with the login code to the e-mail address. The user must then send this code in a second request to /self-service/login/api in order to log in.

What are appropriate HTTP status codes to return by each of the requests, and why?

> <html>
> <body>
> <p>For the two requests you described, here are appropriate HTTP status codes and explanations for each:</p><h3>1. Sending the Login Code Request (POST <code>/self-service/login/api</code>)</h3><p>When a user sends their email address to request a login code, you could use one of the following status codes:</p><ul><li><p><strong><code>200 OK</code></strong>: This indicates that the request was successfully processed. Even if the email address does not exist in your system, returning <code>200 OK</code> helps avoid leaking information about which email addresses are registered and which are not. You can provide a generic message like "If the email address exists, a login code has been sent."</p></li><li><p><strong><code>202 Accepted</code></strong>: This status code implies that the request has been received and will be processed, but the processing is not complete. This is suitable if sending the email code happens asynchronously and may take a short while to complete. The server would return <code>202</code> immediately, while the email is being sent in the background.</p></li><li><p><strong><code>400 Bad Request</code></strong>: Use this if the email address is missing or does not conform to a valid email format. It indicates that there is an issue with the data provided by the user, and the request cannot be processed until it is corrected.</p></li></ul><h3>2. Submitting the Login Code to Authenticate (POST <code>/self-service/login/api</code>)</h3><p>When the user submits the login code, the server needs to validate it. The appropriate status codes could be:</p><ul><li><p><strong><code>200 OK</code></strong>: The login code is valid, and the user has been successfully authenticated. The response may contain details about the session or authentication token.</p></li><li><p><strong><code>400 Bad Request</code></strong>: The login code is missing or does not meet the required format. This indicates a client-side error, where the provided data cannot be processed correctly.</p></li><li><p><strong><code>401 Unauthorized</code></strong>: The login code is incorrect or expired. This means the user was not authenticated successfully because the provided credentials were invalid.</p></li><li><p><strong><code>403 Forbidden</code></strong>: If, for some reason, the user is not allowed to log in (e.g., the account is blocked or not active), this status code can be used to indicate that the request is understood, but the action is forbidden.</p></li><li><p><strong><code>429 Too Many Requests</code></strong>: Use this status code if the user attempts to log in too many times with incorrect codes within a short period. This helps prevent brute force attacks by enforcing rate limiting.</p></li></ul><h3>Summary</h3>

| Scenario                                           | Endpoint                      | HTTP Status Code       | Explanation                                                                                               |
|----------------------------------------------------|------------------------------|------------------------|-----------------------------------------------------------------------------------------------------------|
| User requests a login code                         | `POST /self-service/login/api` | `200 OK`               | Request processed successfully. Does not reveal if the email exists.                                      |
|                                                    |                              | `202 Accepted`         | Request accepted for asynchronous processing (email sending in the background).                           |
|                                                    |                              | `400 Bad Request`      | Invalid email format or missing email field.                                                              |
| User submits the login code                        | `POST /self-service/login/api` | `200 OK`               | Login code is correct, and the user is authenticated.                                                     |
|                                                    |                              | `400 Bad Request`      | Login code is missing or invalidly formatted.                                                             |
|                                                    |                              | `401 Unauthorized`     | Login code is incorrect or expired.                                                                       |
|                                                    |                              | `403 Forbidden`        | User is blocked or not allowed to log in for other reasons.                                               |
|                                                    |                              | `429 Too Many Requests`| User has made too many failed attempts in a short time frame. Rate limiting.                              |
> <p>This setup provides clear, standard HTTP responses that appropriately handle success, client errors, and security concerns.</p>
> </body>
> </html>

Some developer argues that the first request should return 400 even if the request was valid and the email was sent. What do you think about it?

> <div class="markdown prose w-full break-words dark:prose-invert dark"><p>Returning a <strong><code>400 Bad Request</code></strong> even if the request was valid and the email was successfully sent would not be an appropriate use of HTTP status codes. Here’s why:</p><h3>Reasoning Against Using <code>400 Bad Request</code> for a Valid Request</h3><ol><li><p><strong>Semantic Meaning of <code>400 Bad Request</code></strong>:
> The <strong><code>400 Bad Request</code></strong> status code indicates that the server could not understand the request due to invalid syntax, missing data, or incorrect data formats. This is a clear indication that there was something wrong with the client's request, not with the processing on the server. Returning <code>400</code> would mislead the client into thinking they made a mistake, even when the request was properly formed and processed.</p></li><li><p><strong>HTTP Standards and Best Practices</strong>:
> According to the HTTP specification, status codes in the <code>4xx</code> range are client errors, meaning that the client has made a mistake in their request. Since the client request is valid (e.g., correctly formatted email, proper JSON structure), responding with <code>400</code> would violate this principle. It would be an anti-pattern because you're signaling an error when there isn’t one.</p></li><li><p><strong>Developer Experience</strong>:
> For developers integrating with your API, seeing a <strong><code>400 Bad Request</code></strong> would lead them to assume there’s something wrong with how they’re sending the email address. They may waste time troubleshooting an issue that doesn’t exist, leading to frustration and confusion. Properly using <strong><code>200 OK</code></strong> or <strong><code>202 Accepted</code></strong> indicates that the request was successful, even if there’s no explicit acknowledgment about whether an email was sent.</p></li><li><p><strong>Avoiding Information Disclosure</strong>:
> The suggested use of <code>400</code> might be to prevent disclosing whether an email exists in the system, but this can be achieved without violating status code conventions. You can still return <strong><code>200 OK</code></strong> with a generic message ("If the email exists, a code has been sent") without revealing if the email is in the database.</p></li></ol><h3>Preferred Approach</h3><p>Given these considerations, the appropriate behavior would be:</p><ul><li>Return <strong><code>200 OK</code></strong> or <strong><code>202 Accepted</code></strong> for valid requests, indicating the request was processed.</li><li>Use <strong><code>400 Bad Request</code></strong> only when the request is invalid due to missing or malformed data.</li><li>Ensure responses are generic to avoid disclosing sensitive information, such as whether an email address is registered.</li></ul><h3>Example Response</h3><p>For a valid request:</p><ul><li><strong>HTTP Status</strong>: <code>200 OK</code></li><li><strong>Response Body</strong>: <code>{ "message": "If the email exists, a login code has been sent." }</code></li></ul><p>For an invalid request (e.g., malformed email):</p><ul><li><strong>HTTP Status</strong>: <code>400 Bad Request</code></li><li><strong>Response Body</strong>: <code>{ "error": "Invalid email address." }</code></li></ul><p>This approach aligns with HTTP standards and ensures a clearer, more developer-friendly experience.</p></div>


First you say having to parse the body is bad. When showed to you that this is exactly what's needed *now* but no longer with the suggested solution, you jump to using ChatGPT as an argument. Now that ChatGPT disagrees with you, what's your next argument?

I'm very thankful for the work you're putting into Ory and that it is open source. You're the maintainer and you can do whatever you want. However, the way this real issue is handled is very frustrating and makes me lose trust in this product.

**Comment by optimix** [comment]
I've lost a lot of time because of this 400 code trying to figure out what was wrong on the client side seeking and where is the Bad Request.
It finally turned out that there was no Bad Request after all... only that the registration was not yet complete.

**Comment by aeneasr** [comment]
While this won't be tackled with v1 (due to too many breaking changes), we are planning on refactoring the internal flow model at some point (v2) in which case we can also separate the status codes more granularly. We will still return a non-success code for any form fill that is incomplete or being  processed. 200 OK will only ever be sent when the flow is successfully completed.

For all those posting that you've lost a lot of time due to this, where exactly did you lose the time? We can probably improve the SDK to better handle this status code.

**Comment by optimix** [comment]
Thank you very much for reopening this issue for v2.
I understand that adding breaking changes in v1 is not a good idea because it could break existing usages.

On my side I used the _Registration Flow for Native Apps_ with _@ory/client_ in a native web application using _capacitor_.

When I read the documentation of the API it was not clear to me that the API must be called several times before it returns a 20x code.
Then when I saw the _400 Bad Request_ status code in the dev tools, I spent quite a while to see what was wrong with the request that was sent.
Since I was setting up the registration using an e-mail code only (and couldn't find much examples for my use case), I was unsure what **I** did wrong because the request was considered bad.
But I was lucky enough to find this issue with all the comments and then I understood that the request was actually valid.

Having different return codes looks much better and also allows the developer to take another look at the doc and see the meaning of each one.
* 202 if no specific issue but not yet logged in because flow not finished - this one looks pretty nice
* 200 if flow finished and logged in
* 401 for unauthorized if 2fa looks good as well
* 422 - not sure, but why not ?

400 should probably be used only if the request is really bad and the body JSON is junk ?

**Comment by aeneasr** [comment]
Thanks optimix, we definitely need to improve the docs here as it seems a point of frustration for many. fyi @christiannwamba 

Regarding concrete status codes etc, we will most likely publish a design document prior to implementation for discussion. 400 bad request may be sent on invalid JSON, but can also be send on invalid payloads. We originally got the idea from the IETF published OAuth2 spec (and wanted to keep consistency), where basically all errors are 400 bad request errors if the request parameters do not validate. I think it's common to send a 400 error on payload validation issues (whether that's malformed or invalid data doesn't really matter).

**Comment by Juulaps** [comment]
> I've lost a lot of time because of this 400 code trying to figure out what was wrong on the client side seeking and where is the Bad Request. It finally turned out that there was no Bad Request after all... only that the registration was not yet complete.

Same situation here. We are using a different library for data fetching instead of the Ory SDK and getting a 400 here was definitely unexpected.

**Comment by christiannwamba** [comment]
Tracking the issue now in the docs repo so I can improve/clarify

**Comment by dkasenka** [comment]
> Thanks optimix, we definitely need to improve the docs here as it seems a point of frustration for many. fyi [@christiannwamba](https://github.com/christiannwamba)
> 
> Regarding concrete status codes etc, we will most likely publish a design document prior to implementation for discussion. 400 bad request may be sent on invalid JSON, but can also be send on invalid payloads. We originally got the idea from the IETF published OAuth2 spec (and wanted to keep consistency), where basically all errors are 400 bad request errors if the request parameters do not validate. I think it's common to send a 400 error on payload validation issues (whether that's malformed or invalid data doesn't really matter).

Hello [aeneasr](https://github.com/aeneasr),
Could you please provide the section in oAuth2.0 spec where it tells that it has to be used 400 instead of 200? 
Specification: https://datatracker.ietf.org/doc/html/rfc6749

**Comment by dkasenka** [comment]
Hello [aeneasr](https://github.com/aeneasr),

Could you please explain why you prefer this way? If it's because of the specification, can you also please provide the section where I can read it, since I haven't found why to use this approach so far?

>We will still return a non-success code for any form fill that is incomplete or being processed. 200 OK will only ever be sent when the flow is successfully completed.

**Comment by aeneasr** [comment]
You will find the answer in the many messages already written in this thread.

**Comment by IMG-AJB** [comment]
> Alternative way of looking at this problem, why is the login API being used for a request that isn't intended to log a user in?

Exactly. This API is overloaded. It should have been broken up into separate endpoints. Not likely to happen now though.

This problem just wasted three hours of my time as well. And now I need to implement logic to inspect the body that shouldn't be necessary.

Writing the extra code to parse the response is fine, but there should be a gotcha warning in the docs that explains this unexpected 400. In my case I was dealing with _real_ 400s and then didn't even realize that I'd solved the problem because I just kept seeing 400, 400, 400.

400 in the case of a successful request to initiate an OTC login is not the correct code, 202 is the correct code, at least in the case of initiating a passwordless OTC login: received the request, processing it (sending the email).

**Comment by LeoVS09** [comment]
Want to give a few cents to increase the priority of this issue. I'm CTO of a startup, and we were considering multiple providers, and Ory Kratos was looking like the best solution before we started implementation. 

It is very sad that Kratos does not follow a regular RESTful API, with login and register methods. This results in much custom code for just simple login and registration. On top of that SDK docs do not clearly explain how to work with this API and erros. You need to first start the authentication flow, then update this flow, then check the error, which can be not an error at all, but successful behavior. On top of the complexity with session management and CSRF token handling, through parsing of some strange `nodes` object.

Overcomplication was acceptable, but till the moment that I realized that incorrect status handling was not a bug, but a design choice, for which no reason was provided. 
This issue has been open for more than a year, and there are still no updates on when it will be resolved in v2, or at least when a configuration option to disable it will be added. 

It means that on top of writing excessive error-handling code in front-end, we will need to adjust our monitoring system, to somehow distinguish correct 400 "Bad Request" from incorrect 400 "Bad Request", and not trigger alarms. Plus, we will need to configure proxies to also find a way to cache this strange API in the future, which I'm not yet sure is possible.

Unfortunately, I will switch to another solution instead of Kratos. Their complexity is at least clearly mentioned in the docs.

**Comment by monkey-programmer** [comment]
I share the same frustration with identity-first login flows. Personally, I don’t think breaking changes are a big issue here, since this could be handled with a simple configuration flag to switch the behavior, right?
